### PR TITLE
Refactor three: Extract deposit sweeping into a separate library

### PR DESCRIPTION
**Pull request by pdyraga**
~~Depends on #198~~

A refactoring of the Bridge contract is required given the contract size
constraints. This change is another step on this journey. Sweep
functionality is extracted to a separate, externally linked library. No
changes to the logic in the sweeping code were applied.
There were to small changes to the code and docs in `BridgeState`
committed separately: 7b4fefd1b40f51e9f49152cbebb95242f37a8efd, 0cb7816ce2f1ff338f3fb72fdfd82920dfb7d9b6.

Bridge contract size comparison before and after:
```
 |  Bridge               ·     24.852   │
 |  Bridge               ·     20.689   │
```

Gas costs comparison before and after:
```
 |  revealDeposit           ·    94028  ·    96659  ·    94690   │
 |  revealDeposit           ·    96041  ·    98672  ·    96703   │
 |  submitSweepProof        ·   190576  ·   344943  ·   264072   │
 |  submitSweepProof        ·   196251  ·   350622  ·   269763   │
 |  submitRedemptionProof   ·   126354  ·   203246  ·   171551   │
 |  submitRedemptionProof   ·   126280  ·   203172  ·   171481   │
```

### CODE REVIEW

**Comment by pdyraga** [comment]
Rebased given that #198 has been merged. Undrafting!

File: solidity/contracts/bridge/Bridge.sol [hunk 1]
```diff
@@ -2120,6 +1636,14 @@ contract Bridge is Ownable, EcdsaWalletOwner {
         return (targetWalletsHash, outputsTotalValue);
     }
 
+    /// @return bank Address of the Bank the Bridge belongs to.
+    /// @return relay Address of the Bitcoin relay providing the current Bitcoin
+    ///         network difficulty.
+    function getContracts() external view returns (Bank bank, IRelay relay) {
```
Comment (lukasz-zimnoch) [inline solidity/contracts/bridge/Bridge.sol:1642]: Non-blocking but maybe `getPeerContracts` or something similar? `getContracts` looks very generic IMHO.

File: solidity/contracts/bridge/Bridge.sol [hunk 2]
```diff
@@ -2133,13 +1657,30 @@ contract Bridge is Ownable, EcdsaWalletOwner {
...
+    ///         Bitcoin chain required to successfully evaluate an SPV proof.
     function depositParameters()
         external
         view
-        returns (uint64 depositDustThreshold, uint64 depositTreasuryFeeDivisor)
+        returns (
+            uint64 depositDustThreshold,
+            uint64 depositTreasuryFeeDivisor,
+            uint64 depositTxMaxFee,
+            address treasury,
```
Comment (lukasz-zimnoch) [inline solidity/contracts/bridge/Bridge.sol:1675]: The `treasury` is also used for redemptions. Shouldn't it live under `getContracts`? I'm aware the treasury will not be necessarily a contract but it seems to be something like "an address of external dependency" not just a "deposit parameter".

File: solidity/contracts/bridge/Sweep.sol [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+
+import {BTCUtils} from "@keep-network/bitcoin-spv-sol/contracts/BTCUtils.sol";
+
+import "./BitcoinTx.sol";
+import "./BridgeState.sol";
+import "./Wallets.sol";
+
+import "../bank/Bank.sol";
+
+library Sweep {
```
Comment (lukasz-zimnoch) [inline solidity/contracts/bridge/Sweep.sol:26]: Non-blocking: A `@notice` explaining the purpose can be a good idea here.

File: solidity/contracts/bridge/Sweep.sol [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+
+    /// @notice Resolves sweeping wallet based on the provided wallet public key
+    ///         hash. Validates the wallet state and current main UTXO, as
+    ///         currently known on the Ethereum chain.
+    /// @param walletPubKeyHash public key hash of the wallet proving the sweep
+    ///        Bitcoin transaction.
+    /// @param mainUtxo Data of the wallet's main UTXO, as currently known on
+    ///        the Ethereum chain. If no main UTXO exists for the given wallet,
+    ///        this parameter is ignored
+    function resolveWallet(
```
Comment (lukasz-zimnoch) [inline solidity/contracts/bridge/Sweep.sol:206]: Haven't we just renamed it in https://github.com/keep-network/tbtc-v2/pull/198?

File: solidity/hardhat.config.ts [hunk 5]
```diff
@@ -101,7 +101,7 @@ const config: HardhatUserConfig = {
     disambiguatePaths: false,
     runOnCompile: true,
     strict: true,
-    except: ["BridgeStub$", "Bridge$"], // TODO: Remove it once the problem of contracts' size is solved
+    except: ["BridgeStub$"], // TODO: Remove it once the problem of contracts' size is solved
```
Comment (lukasz-zimnoch) [inline solidity/hardhat.config.ts:104]: 🎉 🎉 🎉

File: solidity/contracts/bridge/Sweep.sol (hunk 4 above)
Comment (pdyraga) [reply solidity/contracts/bridge/Sweep.sol:206]: Yeah, it was lost during the rebase. I'll fix it, good catch!

File: solidity/contracts/bridge/Bridge.sol (hunk 2 above)
Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol:1675]: Let's talk about it in #206. I extracted it there into a separate function but I can also move it to `getContracts`. The thing is that treasury is not necessarily a contract.

File: solidity/contracts/bridge/Sweep.sol (hunk 3 above)
Comment (pdyraga) [reply solidity/contracts/bridge/Sweep.sol:26]: Yep, I'll add it in a final cleanup PR. Let's keep this discussion unresolved until then.

File: solidity/contracts/bridge/Sweep.sol (hunk 4 above)
Comment (pdyraga) [reply solidity/contracts/bridge/Sweep.sol]: https://github.com/keep-network/tbtc-v2/pull/200/commits/3f2279f4ff608aa2e588d5388b2e51e4a9226ee1

File: solidity/contracts/bridge/BridgeState.sol [hunk 6]
```diff
@@ -50,5 +75,29 @@ library BridgeState {
...
+        returns (BitcoinTx.ProofDifficulty memory proofDifficulty)
+    {
+        IRelay relay = self.relay;
+        proofDifficulty.currentEpochDifficulty = relay
+            .getCurrentEpochDifficulty();
+        proofDifficulty.previousEpochDifficulty = relay
+            .getPrevEpochDifficulty();
+        proofDifficulty.difficultyFactor = self.txProofDifficultyFactor;
+
+        return proofDifficulty;
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/BridgeState.sol:102]: If we move it to the `Bridge` contract then we need to call it from there and pass `BitcoinTx.ProofDifficulty` as a parameter to sweep/redeem/move funds library functions. This is possible but it's another variable on the stack and EVM may bomb us with "stack too deep" :/ 

Let's maybe keep this discussion as unresolved and try to move it at the end once we'll know if "stack too deep" is a problem or not? I've added a TODO in https://github.com/keep-network/tbtc-v2/pull/200/commits/c3a6dbd1f65ff8b230fb546f19affee7ac507890

[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/BridgeState.sol:102]: I do agree with your comment though.

**Review by lukasz-zimnoch — APPROVED** [review]

File: solidity/contracts/bridge/Bridge.sol (hunk 1 above)
Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol:1642]: Yeah, let's revisit in the final cleanup PR. Maybe we will just expose them in separate functions, we have a plenty of space in Bridge now.

File: solidity/contracts/bridge/BridgeState.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/BridgeState.sol:102]: Problem solved in https://github.com/keep-network/tbtc-v2/pull/211

**Pull request merged**